src/client/QXmppHttpUploadManager.cpp

```diff
@@ -0,0 +1,372 @@
+#include <QPointer>
+
+using namespace QXmpp::Private;
+
+struct QXmppHttpUploadPrivate
+{
+    explicit QXmppHttpUploadPrivate(QXmppHttpUpload *q) : q(q) {}
+
+    QUrl getUrl;
+    std::optional<QXmppError> error;
+    quint64 bytesSent = 0;
+    quint64 bytesTotal = 0;
+    QPointer<QNetworkReply> reply;
+    bool finished = false;
+    bool cancelled = false;
+
+    void reportError(QXmppStanza::Error err)
+    {
+        error = QXmppError {
+            err.text(),
+            std::move(err)
+        };
+    }
+    void reportError(QXmppError newError)
+    {
+        error.emplace(std::move(newError));
+    }
+    void reportFinished()
+    {
+        if (!finished) {
+            finished = true;
+            emit q->finished(result());
+        }
+    }
+    void reportProgress(quint64 sent, quint64 total)
+    {
+        if (total == 0 && bytesTotal > 0) {
+            // QNetworkReply resets the progress in the end
+            // ignore that so we can still see how large the file was
+            return;
+        }
+
+        if (bytesSent != sent || bytesTotal != total) {
+            bytesSent = sent;
+            bytesTotal = total;
+            emit q->progressChanged();
+        }
+    }
+    [[nodiscard]] QXmppHttpUpload::Result result() const
+    {
+        // assumes finished = true
+        if (error) {
+            return *error;
+        }
+        if (cancelled) {
+            return QXmppHttpUpload::Cancelled();
+        }
+        return getUrl;
+    }
+
+private:
+    QXmppHttpUpload *q;
+};
+
+///
+/// \class QXmppHttpUpload
+///
+/// Object that represents an ongoing or finished upload.
+///
+
+///
+/// \property QXmppHttpUpload::progress
+///
+/// the current progress of the upload as a floating point number between 0 and 1
+///
+
+///
+/// \property QXmppHttpUpload::bytesSent
+///
+/// Number of bytes sent so far
+///
+
+///
+/// \property QXmppHttpUpload::bytesTotal
+///
+/// Number of bytes that need to be sent in total to complete the upload
+///
+
+///
+/// \fn QXmppHttpUpload::progressChanged
+///
+/// Emitted when the upload has made progress.
+///
+
+///
+/// \fn QXmppHttpUpload::finished
+///
+/// Emitted when the upload has finished for any reason (success, cancelled, error).
+///
+/// \param result Result of the upload
+///
+
+QXmppHttpUpload::~QXmppHttpUpload() = default;
+
+///
+/// \class QXmppHttpUploadManager
+///
+/// The upload manager allows to upload a file to a server via HTTP File Upload (XEP-363).
+/// This can be used for sending files to other users.
+///
+/// You can enable it using
+/// \code
+/// client.addNewExtension<QXmppUploadRequestManager>();
+/// auto *uploadManager = client.addNewExtension<QXmppHttpUploadManager>();
+/// \endcode
+///
+
+///
+/// \typedef QXmppHttpUpload::Result
+///
+/// Represents the result of an upload.
+/// It can either be a an url, to the uploaded file,
+/// a QXmppHttpUpload::Cancelled unit struct, or an error as QXmppError.
+///
```

Comment:
```suggestion
/// Represents the result of an upload. It can either be a an url, to the
/// uploaded file, a QXmppHttpUpload::Cancelled unit struct, or an error as
/// QXmppError.
///
```